### PR TITLE
[ML-DSA] Add pre-hashed API & domain separation

### LINKED ISSUE #500: Pre-Hash Signing and Verification variants

### LINKED ISSUE #501: Domain separation between pre-hash and hashed signing / verification

### LINKED ISSUE #607: Re-enable (updated) Wycheproof tests once all changes are addressed

### PR DESCRIPTION
**Pull request by jschneider-bensch**
This PR introduces a new pre-hashed API for ML-DSA and the additional domain separation via the `ctx` parameter in signing and verification.

The standard says that any NIST-approved hash function or XOF can be used to perform the pre-hashing in the HashML-DSA variant and explicitly lists cases for SHA-256, SHA-512, and SHAKE-128 in Algorithms 4 and 5. From this it wasn't clear to me if these three are mandatory for implementing the standard or if e.g. one of them is sufficient. For the sake of keeping things simple and additional dependencies in the crate minimal I've chosen to implement HashML-DSA just for SHAKE-128 pre-hashing, but we should be able to extend it to other NIST-approved hash functions or XOFs if desired, hopefully without much effort.

With these changes, we test successfully against the Wycheproof vectors for the final standard version, giving some confidence that we've implemented all changes which affect signing and verification. I've thus re-enabled these tests and extended them somewhat such that we test all variants (excluding pre-hashed) in all optimizations.

Fixes #500 
Fixes #501
Fixes #607

### CODE REVIEW

**Comment by jschneider-bensch** [comment]
There are no Wycheproof vectors for the pre-hashed variant, but I've modified the spec-like python version we use to generate our own KATs so we can generate KATs for the pre-hashed version as well.

File: libcrux-ml-dsa/tests/wycheproof_sign.rs [hunk 1]
```diff
@@ -19,7 +19,7 @@ macro_rules! wycheproof_sign_test {
         fn $name() {
             let katfile_path = Path::new("tests")
                 .join("wycheproof")
-                .join(format!("mldsa_{}_draft_sign_test.json", $parameter_set));
+                .join(format!("mldsa_{}_standard_sign_test.json", $parameter_set));
```
Comment (franziskuskiefer) [inline libcrux-ml-dsa/tests/wycheproof_sign.rs:22]: Please add a comment or readme with the commit in wycheproof where these files come from.

**Review by franziskuskiefer — CHANGES_REQUESTED** [review]
Thanks, lgtm with some nits.

File: libcrux-ml-dsa/src/ml_dsa_generic.rs [hunk 2]
```diff
@@ -394,3 +522,144 @@ pub(crate) fn verify<
...
+    context: &[u8],
+    pre_hash_oid: Option<&PreHashOID>,
+) -> ([u8; CONTEXT_MAX_LEN + PRE_HASH_OID_LEN + 2], usize) {
+    debug_assert!(context.len() <= CONTEXT_MAX_LEN);
+    let mut domain_separated_context = [0u8; CONTEXT_MAX_LEN + PRE_HASH_OID_LEN + 2];
+    domain_separated_context[1] = context.len() as u8;
+    let mut len = context.len() + 2;
+    domain_separated_context[2..len].copy_from_slice(context);
+
+    if let Some(pre_hash_oid) = pre_hash_oid {
```
[the remark this answers is not in the document]
Comment (jschneider-bensch) [reply libcrux-ml-dsa/src/ml_dsa_generic.rs:657]: You mean always return a fixed size array of `CONTEXT_MAX_LEN + PRE_HASH_OID_LEN + 2` bytes and get the actually needed length of it in a separate function?

[the remark this answers is not in the document]
Comment (franziskuskiefer) [reply libcrux-ml-dsa/src/ml_dsa_generic.rs:657]: No, always return something with the correct length. If you don't use the same function for regular and pre-hased, that should be doable I think?

[the remark this answers is not in the document]
Comment (jschneider-bensch) [reply libcrux-ml-dsa/src/ml_dsa_generic.rs:657]: I see! But `context` is variable-length (<= 255 bytes) so we can't know the size statically, or am I missing something?

File: libcrux-ml-dsa/src/ml_dsa_generic.rs [hunk 3]
```diff
@@ -100,11 +101,76 @@ pub enum VerificationError {
...
+>(
+    signing_key: &[u8; SIGNING_KEY_SIZE],
+    message: &[u8],
+    context: &[u8],
+    randomness: [u8; SIGNING_RANDOMNESS_SIZE],
+) -> Result<MLDSASignature<SIGNATURE_SIZE>, SigningError> {
+    if context.len() > CONTEXT_MAX_LEN {
+        Err(SigningError::ContextTooLongError)
+    } else {
+        let (domain_separated_context, ctx_len) =
```
[the remark this answers is not in the document]
Comment (jschneider-bensch) [reply libcrux-ml-dsa/src/ml_dsa_generic.rs:144]: Indeed, amazing!

**Review by franziskuskiefer — APPROVED** [review]
Leaving an r+ assuming that the domain separation gets resolved as discussed offline.

File: libcrux-ml-dsa/src/ml_dsa_generic.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jschneider-bensch) [reply libcrux-ml-dsa/src/ml_dsa_generic.rs]: I've implemented the domain separation without copying as we discussed offline.

**Pull request merged**